components/feature/p2p/src/main/java/mozilla/components/feature/p2p/internal/P2PController.kt

```diff
@@ -0,0 +1,232 @@
+                    is ConnectionState.ReadyToSend -> R.string.mozac_feature_p2p_connected
+                    is ConnectionState.Sending -> R.string.mozac_feature_p2p_sending
+                    is ConnectionState.Failure -> R.string.mozac_feature_p2p_failure
+                }
+            )
+        }
+
+        override fun onMessageDelivered(payloadId: Long) {
+            outgoingMessages.get(payloadId)?.let {
+                view.reportSendComplete(
+                    if (it == URL_INDICATOR) {
+                        R.string.mozac_feature_p2p_url_sent
+                    } else {
+                        R.string.mozac_feature_p2p_page_sent
+                    }
+                )
+                // Is it better to remove entries for delivered messages from the map
+                // or to leave them in (which is more functional-style)?
+                outgoingMessages.remove(payloadId)
+            } ?: run {
+                logger.error("Sent message id was not recognized")
+            }
+        }
+
+        override fun onMessageReceived(neighborId: String, neighborName: String?, message: String) {
+            if (message.length > 1) {
+                when (message[0]) {
+                    HTML_INDICATOR -> view.receivePage(
+                        neighborId, neighborName, message.substring(1)
+                    )
+                    URL_INDICATOR -> view.receiveUrl(
+                        neighborId, neighborName, message.substring(1)
+                    )
+                    else -> reportError("Cannot parse incoming message $message")
+                }
+            } else {
+                reportError("Trivial message received: '$message'")
+            }
+        }
+    }
+
+    fun start() {
+        view.listener = this
+        if (nearbyConnection == null || savedConnectionState == null) {
+            nearbyConnection = thunk()
+            savedConnectionState = null
+        } else {
+            // If a connection already existed, update the UI to reflect its status.
+            when (savedConnectionState) {
+                is ConnectionState.Isolated -> view.initializeButtons(true, false)
+                is ConnectionState.Advertising,
+                is ConnectionState.Discovering,
+                is ConnectionState.Initiating,
+                is ConnectionState.Authenticating,
+                is ConnectionState.Connecting -> view.initializeButtons(false, false)
+                is ConnectionState.ReadyToSend -> view.initializeButtons(false, true)
+                is ConnectionState.Failure -> view.initializeButtons(false, false)
+            }
+        }
+        nearbyConnection?.register(observer, view as P2PBar)
+    }
+
+    fun stop() {
+        nearbyConnection?.unregisterObservers()
+    }
+
+    @Synchronized
+    private fun reportError(msg: String) {
+        Logger.error(msg)
+        view.failure(msg)
+    }
+
+    private inline fun <reified T : ConnectionState> cast() =
```

Comment:
Having to do this cast doesn't feel great. 😐  I think we need to re-think adding functionality to a state object.

Reply:
We could eliminate the `cast` function, but then we'd have to do a lot of `as?` casts. Would I be correct in assuming that's just as bad?

Reply:
It's fine. I think it's the `cast<ConnectionState.Authenticating>()?.accept()` state calls that require the cast in order to invoke feel awkward to have to use, but I have no good suggestions right now. 🤷‍♂